File: packages/mda-config/python/src/snoai_mda_config/requires.py
"""MDA §10-3.3 / §10-4 requires.network enforcement."""

from __future__ import annotations

import ipaddress
import re
from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import Any, cast

from .errors import ErrorCategory, MdaConfigError

_PRIVATE_V4_NETWORKS: tuple[ipaddress.IPv4Network, ...] = (
    ipaddress.IPv4Network("10.0.0.0/8"),
    ipaddress.IPv4Network("172.16.0.0/12"),
    ipaddress.IPv4Network("192.168.0.0/16"),
)
_PRIVATE_V6_NETWORKS: tuple[ipaddress.IPv6Network, ...] = (
    ipaddress.IPv6Network("fc00::/7"),
)


@dataclass(frozen=True)
class RequiresEnvironment:
    """Operator-supplied environment for §10-4 enforcement."""

    allowed_networks: Sequence[str] = field(default_factory=tuple)


def enforce_requires(requires: dict[str, Any] | None, env: RequiresEnvironment) -> None:
    """MDA §10-4 enforces the top-level requires.network value."""

    if requires is None:
        return
    if "network" not in requires:
        return

    network = requires["network"]
    allowed = set(env.allowed_networks)
    wildcard = "*" in allowed

    if network == "none":
        return

    if network == "local":
        for host in allowed:
            if not _is_local_host(host):
                raise MdaConfigError(
                    ErrorCategory.RequiresNotSatisfied,
                    "requires.network=local but operator permits non-local host",
                    {"key": "network", "host": host},
                )
        return

    if network == "public":
        if wildcard:
            return
        raise MdaConfigError(
            ErrorCategory.RequiresNotSatisfied,
            "requires.network=public but operator does not grant wildcard '*'",
            {"key": "network"},
        )

    if isinstance(network, list):
        network_items = cast("list[Any]", network)
        for host_candidate in network_items:
            if not isinstance(host_candidate, str) or not host_candidate:
                _invalid_network_shape(network)
        network_hosts = cast("list[str]", network_items)
        if wildcard:
            return
        for host in network_hosts:
            if not _is_network_allowed(host, allowed):
                raise MdaConfigError(
                    ErrorCategory.RequiresNotSatisfied,
                    f"requires.network host '{host}' not in operator allow-list",
                    {"key": "network", "host": host, "allowed": sorted(allowed)},
                )
        return

    _invalid_network_shape(network)


def _invalid_network_shape(value: Any) -> None:
    raise MdaConfigError(
        ErrorCategory.RequiresNotSatisfied,
        "requires.network has an invalid shape",
        {"key": "network", "reason": "invalid-shape", "got": value},
    )


def _is_network_allowed(required_host: str, allowed: set[str]) -> bool:
    if required_host in allowed:
        return True
    return any(_host_matches_pattern(required_host, pattern) for pattern in allowed)


def _host_matches_pattern(host: str, pattern: str) -> bool:
    if "*" not in pattern:
        return host == pattern
    escaped = re.escape(pattern).replace(r"\*", "[^.]+")
    return re.fullmatch(escaped, host) is not None


def _is_local_host(host: str) -> bool:
    normalized = host.rstrip(".").lower()
    if (
        normalized == "localhost"
        or normalized.endswith(".localhost")
        or normalized.endswith(".local")
        or normalized.endswith(".internal")
    ):
        return True
    try:
        address = ipaddress.ip_address(normalized)
    except ValueError:
        return False
    if address.is_loopback:
        return True
    if isinstance(address, ipaddress.IPv4Address):
        return any(address in network for network in _PRIVATE_V4_NETWORKS)
    return any(address in network for network in _PRIVATE_V6_NETWORKS)

File: packages/mda-config/rust/src/requires.rs
use std::net::IpAddr;

use regex::Regex;
use serde_json::Value;

use crate::errors::{ErrorCategory, MdaConfigError, Result};

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RequiresEnvironment {
    pub allowed_networks: Vec<String>,
}

pub fn enforce_requires(requires: Option<&Value>, env: &RequiresEnvironment) -> Result<()> {
    let Some(requires) = requires else {
        return Ok(());
    };
    let Some(network) = requires.get("network") else {
        return Ok(());
    };

    let allowed: std::collections::HashSet<&str> =
        env.allowed_networks.iter().map(String::as_str).collect();
    let wildcard = allowed.contains("*");

    if network == "none" {
        return Ok(());
    }

    if network == "local" {
        for host in &env.allowed_networks {
            if !is_local_host(host) {
                return Err(MdaConfigError::with_details(
                    ErrorCategory::RequiresNotSatisfied,
                    "requires.network=local but operator permits non-local host",
                    serde_json::json!({ "key": "network", "host": host }),
                ));
            }
        }
        return Ok(());
    }

    if network == "public" {
        if wildcard {
            return Ok(());
        }
        return Err(MdaConfigError::with_details(
            ErrorCategory::RequiresNotSatisfied,
            "requires.network=public but operator does not grant wildcard '*'",
            serde_json::json!({ "key": "network" }),
        ));
    }

    let Some(hosts) = network.as_array() else {
        return invalid_network_shape(network);
    };
    if hosts
        .iter()
        .any(|host| host.as_str().is_none_or(|s| s.is_empty()))
    {
        return invalid_network_shape(network);
    }
    if wildcard {
        return Ok(());
    }
    for host in hosts.iter().filter_map(Value::as_str) {
        if !is_network_allowed(host, &allowed) {
            return Err(MdaConfigError::with_details(
                ErrorCategory::RequiresNotSatisfied,
                format!("requires.network host '{host}' not in operator allow-list"),
                serde_json::json!({
                    "key": "network",
                    "host": host,
                    "allowed": env.allowed_networks,
                }),
            ));
        }
    }
    Ok(())
}

fn invalid_network_shape(value: &Value) -> Result<()> {
    Err(MdaConfigError::with_details(
        ErrorCategory::RequiresNotSatisfied,
        "requires.network has an invalid shape",
        serde_json::json!({ "key": "network", "reason": "invalid-shape", "got": value }),
    ))
}

fn is_network_allowed(host: &str, allowed: &std::collections::HashSet<&str>) -> bool {
    if allowed.contains(host) {
        return true;
    }
    allowed
        .iter()
        .any(|pattern| host_matches_pattern(host, pattern))
}

fn host_matches_pattern(host: &str, pattern: &str) -> bool {
    if !pattern.contains('*') {
        return host == pattern;
    }
    let escaped = regex::escape(pattern).replace("\\*", "[^.]+");
    Regex::new(&format!("^{escaped}$"))
        .map(|re| re.is_match(host))
        .unwrap_or(false)
}

fn is_local_host(host: &str) -> bool {
    let host = host.trim_end_matches('.').to_ascii_lowercase();
    if host == "localhost"
        || host.ends_with(".localhost")
        || host.ends_with(".local")
        || host.ends_with(".internal")
    {
        return true;
    }
    match host.parse::<IpAddr>() {
        Ok(IpAddr::V4(addr)) => addr.is_loopback() || addr.is_private(),
        Ok(IpAddr::V6(addr)) => addr.is_loopback() || (addr.segments()[0] & 0xfe00) == 0xfc00,
        Err(_) => false,
    }
}
